relaynet-cogrpc-jvm/src/main/java/tech/relaycorp/relaynet/cogrpc/client/CogRPCClient.kt

```diff
@@ -1,19 +1,132 @@
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+import tech.relaycorp.relaynet.cogrpc.Authorization
+import tech.relaycorp.relaynet.cogrpc.CargoDelivery
+import tech.relaycorp.relaynet.cogrpc.CargoDeliveryAck
+import tech.relaycorp.relaynet.cogrpc.CargoRelayGrpc
 import tech.relaycorp.relaynet.cogrpc.CogRPC
+import java.net.InetSocketAddress
+import java.net.URL
+import java.util.logging.Level
+import java.util.logging.Logger
 
-interface CogRPCClient {
+class CogRPCClient(
+    serverAddress: String,
+    useTls: Boolean = true
+) {
+
+    private val channel by lazy {
+        val url = URL(serverAddress)
+        val address = InetSocketAddress(
+            url.host,
+            url.port.let { if (it != -1) it else DEFAULT_PORT }
+        )
+        NettyChannelBuilder
+            .forAddress(address)
+            .useTransportSecurity()
+            .run { if (useTls) useTransportSecurity() else usePlaintext() }
+            .build()
+    }
 
     @Throws(Exception::class)
-    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck>
+    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck> {
+        if (cargoes.none()) return emptyFlow()
+
+        val cargoesToAck = cargoes.map { it.localId }.toMutableList()
+        val ackChannel = BroadcastChannel<CogRPC.MessageDeliveryAck>(1)
+        val ackObserver = object : StreamObserver<CargoDeliveryAck> {
+            override fun onNext(value: CargoDeliveryAck) {
+                logger.info("deliverCargo ack ${value.id}")
+                ackChannel.sendBlocking(CogRPC.MessageDeliveryAck(value))
+                cargoesToAck.remove(value.id)
+                if (cargoesToAck.isEmpty()) {
+                    logger.info("deliverCargo ack complete")
+                    ackChannel.close()
+                }
+            }
+
+            override fun onError(t: Throwable) {
+                logger.log(Level.WARNING, "deliverCargo ack error", t)
+                ackChannel.close(t)
+            }
+
+            override fun onCompleted() {
+                logger.info("deliverCargo ack closed")
+                ackChannel.close()
+            }
+        }
+
+        val client = buildClient()
+        val deliveryObserver = client.deliverCargo(ackObserver)
+
+        cargoes.forEach { delivery ->
+            logger.info("deliverCargo next ${delivery.localId}")
+            deliveryObserver.onNext(delivery.toCargoDelivery())
+        }
+        logger.info("deliverCargo complete")
+        deliveryObserver.onCompleted()
+
+        return ackChannel.asFlow()
+    }
 
     @Throws(Exception::class)
-    fun collectCargo(cca: CogRPC.MessageDelivery): Flow<CogRPC.MessageReceived>
+    fun collectCargo(cca: CogRPC.MessageDelivery): Flow<CogRPC.MessageDelivery> {
```

Comment:
Similarly to the comment above, I think the output should be either `Flow<ByteArray>` or `Flow<OutputStream>`. There's no local id at this point, so I don't think we should output one.